app-backend/tile/src/main/scala/LayerCache.scala

```diff
@@ -146,4 +135,69 @@ object LayerCache extends Config with LazyLogging {
         }
       }
     }
+
+  /** Calculate the histogram for the least resolute zoom level to automatically render tiles */
+  def modelLayerGlobalHistogram(
+    ast: MapAlgebraAST,
+    params: EvalParams,
+    toolRun: ToolRun,
+    user: User
+  ): OptionT[Future, Histogram[Double]] =
+    histogramCache.cachingOptionT(s"model-${toolRun.id}-${ast.id}") { implicit ec =>
+      for {
+        lztile <- OptionT(Interpreter.interpretGlobal(ast, params.sources, TileSources.cachedGlobalSource).map(_.toOption))
+        tile   <- OptionT.fromOption[Future](lztile.evaluateDouble)
+      } yield {
+        val hist = StreamingHistogram.fromTile(tile)
+        val currentMetadata = params.metadata.getOrElse(ast.id, NodeMetadata())
+        val updatedMetadata = currentMetadata.copy(histogram = Some(hist))
+        val updatedParams = params.copy(metadata=params.metadata + (ast.id -> updatedMetadata))
+        val updatedToolRun = toolRun.copy(executionParameters=updatedParams.asJson)
+        database.db.run { ToolRuns.updateToolRun(updatedToolRun, toolRun.id, user) }
+
+        hist
+      }
+    }
+
+  /** Calculate all of the prerequisites to evaluation of an AST over a set of tile sources */
+  def toolEvalRequirements(
+    toolRunId: UUID,
+    subNode: Option[UUID],
+    user: User
+  ): OptionT[Future, (ToolRun, Tool.WithRelated, MapAlgebraAST, EvalParams, ColorMap)] =
+    astCache.cachingOptionT(s"tool+run-$toolRunId-${subNode}-${user.id}") { implicit ec =>
+      for {
+        toolRun <- OptionT(database.db.run(ToolRuns.getToolRun(toolRunId, user)))
+        tool    <- OptionT(Tools.getTool(toolRun.tool, user))
+        ast     <- OptionT.fromOption[Future]({
+                     logger.debug(s"Parsing Tool AST with ${tool.definition}")
+                     val entireAST = parseOrThrow[MapAlgebraAST](tool.definition)
+                     subNode.flatMap(id => entireAST.find(id)).orElse(Some(entireAST))
+                   })
+        nodeId  <- OptionT.pure[Future, UUID](subNode.getOrElse(ast.id))
+        params  <- OptionT.pure[Future, EvalParams]({
+                     logger.debug(s"Parsing ToolRun parameters with ${toolRun.executionParameters}")
+                     val parsedParams = parseOrThrow[EvalParams](toolRun.executionParameters)
+                     val md = (parsedParams.metadata.get(ast.id), ast.metadata) match {
```

Comment:
I'm not entirely opposed, though I like how explicit and readable the match is here